fix(pkg/plugin/installer): unix-style paths in HELM_PLUGINS

xodmqjw opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:
Supports unix-style paths (with a colon) in the HELM_PLUGINS env for installing new plugins.

Closes #11310

Signed-off-by: Sergei Karpov <xodmqjw@gmail.com>

**Special notes for your reviewer**:


**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

Commit:
fix(pkg/plugin/installer): unix-style paths in HELM_PLUGINS for helm install

Supports unix-style paths (with a colon) in the HELM_PLUGINS env for installing new plugins.

Closes #11310

Signed-off-by: Sergei Karpov <xodmqjw@gmail.com>

gjenkins8 commented:
I didn't understand the rational of the change here. IIUC, the change is to allow specifying multiple `HELM_PLUGINS` paths, splitting those plugins directory path(s), but then immediately discarding all/any additional paths.

Would it please be possible to explain why this behavior is wanted? (hopefully I'm not missing something obvious :) ). And also, should we consider updating the docs https://helm.sh/docs/topics/plugins/#environment-variables ?

mattfarina commented on pkg/plugin/installer/base_test.go:
```diff
@@ -32,6 +32,10 @@ func TestPath(t *testing.T) {
 			expectPath:     "/helm/data/plugins/helm-secrets",
+		}, {
+			source:         "https://github.com/jkroepke/helm-secrets",
+			helmPluginsDir: "/helm/data/plugins:/remote/helm/data/plugins",
```
This test is going to fail on Windows which is one of our supported environments. We need to use the correct path list separator for each environment. See https://pkg.go.dev/path/filepath#pkg-constants

xodmqjw replied:
I'm not sure where I have to add docs. How should I add it in the current MR? Could you please give me a link where I can find how to do this?

mattfarina reviewed (changes requested):
Thanks for a PR to fix this issue. I have some feedback.

This isn't just for Unix path lists. It applies to Windows as well. You can have path lists there.

I think the docs need to be updated as well. We want to be explicit in telling people where something will be installed if there are multiple locations.

Commit:
fix(pkg/plugin/installer): unix-style paths in HELM_PLUGINS for helm install

Supports unix-style paths (with a colon) in the HELM_PLUGINS env for installing new plugins.

Closes #11310

Signed-off-by: Sergei Karpov <xodmqjw@gmail.com>

tghartland commented:
@gjenkins8 

> I didn't understand the rational of the change here. IIUC, the change is to allow specifying multiple `HELM_PLUGINS` paths, splitting those plugins directory path(s), but then immediately discarding all/any additional paths.

Helm already supports HELM_PLUGINS path with multiple entries, for _reading_ plugins.
https://github.com/helm/helm/blob/8199db309ae73d745b983fb7ea2bfdd944a834ae/pkg/plugin/plugin.go#L261-L262

But if you set up the path like that, it breaks _writing_ (installing) plugins, as the entire env variable is treated as a single directory path.
Helm creates and installs the plugin in that directory, but it is then impossible for it to read the plugin from that directory, as it gets split.

Full example in #11310.

mattfarina commented on pkg/plugin/installer/base.go:
```diff
@@ -30,9 +30,10 @@ type base struct {
 	return base{
 		Source:           source,
-		PluginsDirectory: settings.PluginsDirectory,
+		PluginsDirectory: pluginsDirs[0],
```
This is almost there. If `settings.PluginsDirectory` is an empty string than calling `pluginsDir[0]` will result in a panic. Can we check the length of the slice and if it's 0 use an empty string for the value.

mattfarina reviewed (changes requested):
This is almost there. There is an edge case that can cause a panic. Can you please fix that?

Commit:
fix(pkg/plugin/installer): unix-style paths in HELM_PLUGINS for helm install

Supports unix-style paths (with a colon) in the HELM_PLUGINS env for installing new plugins.

Closes #11310

Signed-off-by: Sergei Karpov <xodmqjw@gmail.com>
